### PR TITLE
 [sike] Refactor key agreement in TLS 1.3

### PR DESCRIPTION
**Pull request by henrydcase**
Previously there where two methods used for key agreemnt
tls.Conn::generateKeyShare and tls.Conn::deriveDHESecret. Both were
used on client and server side. Boolean flag is used in order to
differentiate between key agreement performed on client and on server
side. Which sucks badly.
In order to implement shared secret agreement with KEM it is better to
add method which implements server specific key agreement and provide
default implementation which reuses tls.Conn::generateKeyShare followed
by tls.Conn::deriveDHESecret.
Now, it is possible for most of the DH-style key agreements to reuse
default implementation and for KEM-style key agreement to provide server
specific implementation.

### CODE REVIEW

File: 13.go [hunk 1]
```diff
@@ -61,33 +61,64 @@ type keySchedule13 struct {
...
+		return nil, keyShare{}, err
+	}
+	return ss, publicKey, nil
 }
 
 // Key Exchange strategies per curve type
-type kexNist struct{}     // Used by NIST curves; P-256, P-384, P-512
-type kexX25519 struct{}   // Used by X25519
-type kexSIDHp503 struct{} // Used by SIDH/P503
+type kexNist struct{ defaultServerKEX }     // Used by NIST curves; P-256, P-384, P-512
```
Comment (alxdavids) [inline 13.go]: Since SIDH is also an acronym and capitalised in the variable name, maybe this should be kexNIST?

File: 13.go [hunk 2]
```diff
@@ -232,14 +252,17 @@ CurvePreferenceLoop:
...
-
 	earlyClientCipher, _ := hs.keySchedule.prepareCipher(secretEarlyClient)
 
-	ecdheSecret := c.deriveDHESecret(ks, privateKey)
-	if ecdheSecret == nil {
-		c.sendAlert(alertIllegalParameter)
-		return errors.New("tls: bad ECDHE client share")
+	if ks.group == 0 {
+		c.sendAlert(alertInternalError)
+		return errors.New("tls: HelloRetryRequest not implemented") // TODO(filippo)
```
Comment (alxdavids) [inline 13.go]: Ancient TODO?

File: 13.go [hunk 3]
```diff
@@ -601,19 +624,28 @@ func prepareDigitallySigned(hash crypto.Hash, context string, data []byte) []byt
...
 func (c *Conn) generateKeyShare(curveID CurveID) ([]byte, keyShare, error) {
-	if val, ok := dhKexStrat[curveID]; ok {
+	if val, ok := kexStrat[curveID]; ok {
 		return val.generate(c, curveID)
 	}
 	return nil, keyShare{}, errors.New("tls: preferredCurves includes unsupported curve")
 }
 
 // DH key agreement. ks stores public key, secretKey stores private key used for ephemeral
 // key agreement. Function returns shared secret in case of success or empty slice otherwise.
```
Comment (alxdavids) [inline 13.go]: now returns error also

File: 13.go [hunk 4]
```diff
@@ -601,19 +624,28 @@ func prepareDigitallySigned(hash crypto.Hash, context string, data []byte) []byt
...
+func (c *Conn) keyAgreementClient(ks keyShare, secretKey []byte) ([]byte, error) {
+	if val, ok := kexStrat[ks.group]; ok {
+		return val.keyAgreementClient(c, ks, secretKey)
 	}
-	return nil
+	return nil, errors.New("tls: unsupported group")
+}
+
+// keyAgreementServer generates ephemeral keypair on the on the server side
+// and then uses 'keyShare' (client public key) to dervie shared secret
```
Comment (alxdavids) [inline 13.go]: nit: dervie -> derive

File: 13.go [hunk 5]
```diff
@@ -989,10 +1021,10 @@ func (hs *clientHandshakeState) doTLS13Handshake() error {
 
 	// 0-RTT is not supported yet, so use an empty PSK.
 	hs.keySchedule.setSecret(nil)
-	ecdheSecret := c.deriveDHESecret(serverHello.keyShare, hs.privateKey)
-	if ecdheSecret == nil {
+	ecdheSecret, err := c.keyAgreementClient(serverHello.keyShare, hs.privateKey)
```
Comment (alxdavids) [inline 13.go]: Nit: change variable name (to remove dhe)?

File: 13.go (hunk 2 above)
Comment (henrydcase) [reply 13.go]: yeah, this was just copy paste from somewhere else. Indeed this is not implemented yet, which is a shame a bit

File: 13.go (hunk 1 above)
Comment (henrydcase) [reply 13.go]: good idea

File: 13.go (hunk 4 above)
Comment (henrydcase) [reply 13.go]: done

File: 13.go (hunk 5 above)
Comment (henrydcase) [reply 13.go]: let's make it ``sharedSecret`` as that's what it is

File: 13.go (hunk 3 above)
Comment (henrydcase) [reply 13.go]: comment above updated

**Pull request merged**